src/contracts/libraries/GPv2Encoding.sol

```diff
@@ -115,48 +128,73 @@ library GPv2Encoding {
         bytes32 r = abi.decode(encodedOrder[140:], (bytes32));
         bytes32 s = abi.decode(encodedOrder[172:], (bytes32));
 
-        order.sellToken = tokens[order.sellTokenIndex];
-        order.buyToken = tokens[order.buyTokenIndex];
+        order.sellToken = tokens[sellTokenIndex];
+        order.sellTokenIndex = sellTokenIndex;
+        order.buyToken = tokens[buyTokenIndex];
+        order.buyTokenIndex = buyTokenIndex;
         order.kind = OrderKind((flags >> ORDER_KIND_BIT) & 0x1);
         order.partiallyFillable =
             (flags >> ORDER_PARTIALLY_FILLABLE_BIT) & 0x01 == 0x01;
 
-        bytes32 orderDigest;
-        bytes32 signingDigest;
-
         // NOTE: In order to avoid a memory allocation per call by using the
         // built-in `abi.encode`, we reuse the memory region reserved by the
-        // caller for the order result as input to compute the hash.
+        // caller for the order result as input to compute the hash, using the
+        // memory slot reserved for the order `owner` for the order type hash
+        // which is required by EIP-712 as a prefix to the order data.
+        // Furthermore structs are not packed in Solidity, and there is the
+        // order type hash prefix followed by the 9 order fields to hash for a
+        // total of `10 * sizeof(uint) = 320` bytes.
+        bytes32 orderDigest;
         // solhint-disable-next-line no-inline-assembly
         assembly {
-            // NOTE: Structs are not packed in Solidity, and there are 9 order
-            // fields to hash for a total of `9 * sizeof(uint) = 288` bytes.
-            orderDigest := keccak256(order, 288)
+            mstore(order, ORDER_TYPE_HASH)
+            orderDigest := keccak256(order, 320)
         }
 
-        // NOTE: Now compute the digest that was used for signing. This is not
-        // the same as the order digest as it includes a message signature
-        // prefix as well as the domain separator. This is done using a scratch
-        // region of Solidity memory past the last allocation (which is stored
-        // at `0x40` memory address).
+        // NOTE: Solidity allocates, but does not free, memory when calling the
+        // ABI encoding methods as well as the `ecrecover` precompile. However,
+        // we can restore the free memory pointer to before we made allocations
+        // to effectively free the memory.
+        // <https://solidity.readthedocs.io/en/v0.7.4/internals/layout_in_memory.html>
+        uint256 freeMemoryPointer;
```

Comment:
Overall I like this version much better (good job finding the free memory trick). 
I'd still like to explore the increased memory footprint of storing the hashable order as its own sub-struct. Since the order struct is reused (and could potentially be freed after having parsed all orders) I'd expect the memory footprint to be minor.

Reply:
> explore the increased memory footprint

I think there is negligible increase in the memory footprint. The main issue is that `order.data.field` becomes two `mload`s. We can work around this though, so this might be a worthwhile thing to do!